Update Remix instructions to a working version (#9830)
PostHog/posthog.com: contents/docs/libraries/remix.md

npm install --save posthog-js
```

> First, you'll need to pass the `POSTHOG_API_KEY` to your client through Remix. We recommend placing them on `window.ENV` by following [this guide](https://remix.run/docs/en/main/guides/envvars#browser-environment-variables). You'll need both your API key and instance address (you can find these in your [project settings](https://us.posthog.com/project/settings)).

Reviewer: ```suggestion
First, you'll need to pass both your project API key and instance address from your [project settings](https://us.posthog.com/project/settings) to your client through Remix. We recommend placing the project API key on `window.ENV` like Remix recommends in [this guide](https://remix.run/docs/en/main/guides/envvars#browser-environment-variables).
```

Next, create a new PostHog context in `app/contexts/posthog-context.tsx`. This is necessary because of a [missing export statement](https://github.com/PostHog/posthog-js/issues/908) in `posthog-js`'s `package.json`. 

> ```ts file=app/contexts/posthog-context.tsx

Reviewer: This seems to cause a hydration error and de-opts the code into client side rendering for me. The follow code doesn't cause that problem. What do you think?

```
import posthog from "posthog-js";
import { useEffect, createContext, useContext, useRef } from "react";

type PosthogType = typeof posthog | undefined;

const PosthogContext = createContext<PosthogType>(undefined);

export function PosthogProvider({ children }: { children: React.ReactNode }) {
  const posthogInstanceRef = useRef<PosthogType>(undefined);

  useEffect(() => {

    // Only initialize PostHog once during component mount
    if (posthogInstanceRef.current) return;
    
    const apiKey = window.ENV?.POSTHOG_API_KEY;
    if (!apiKey) return;

    posthogInstanceRef.current = posthog.init(apiKey, {
      api_host: "https://us.i.posthog.com",
      loaded: (posthog) => {
        if (process.env.NODE_ENV === 'development') {
          posthog.debug();
        }
      },
    });
  }, []);

  return (
    <PosthogContext.Provider value={posthogInstanceRef.current}>
      {children}
    </PosthogContext.Provider>
  );
}

export const usePosthog = () => useContext(PosthogContext);
```

Author: Fair enough; React's recommendations around variables that must be initialized only once is pretty bad. 

One call out is that `process.env.NODE_ENV` is not available in the client, so I'll replace it with `window.ENV.IS_PROD` and add a note stating that this assumes the client ENV variables include `IS_PROD`. Note that also if the user follows Remix's recommendations on how to pass env variables to the client, `window.ENV` should always be defined, so we don't need the optional chaining.

Author: Also can you check again and/or push the remix repo you're testing with? I checked again, and I'm not getting the hydration error with the original code (that uses a function and a ref). I'm happy to change it to an effect and state, but given that React's recommendation is to use a function and ref, I'd prefer only changing it to effect and state only if it's not working. 

There's a bunch of different reasons why a hydration error is caused in Remix (including [3rd party browser extensions](https://www.jacobparis.com/content/remix-hydration-errors)); I just want to make sure React's recommendation doesn't work before making the change.

Reviewer: It works fine on the first load, but subsequent loads get the hydration error. Tested on guest Chrome without extensions.

Here's the test repo: https://github.com/ivanagas/remix-cool2

Here's a test site where you can see the error on refresh: https://vercel.com/ivanagas-projects/remix-cool2

Author: Thanks for sharing the repo! Sorry about the delay; I was traveling. Downloaded, built, and ran it both in dev and as production. Unfortunately I haven't been able to replicate the hydration error in either my regular Chrome or incognito; or on the first load, uncached load, or reload: 

<img width="1604" alt="Screenshot 2024-12-10 at 5 02 43 PM" src="https://github.com/user-attachments/assets/ae112c56-65f2-4ae3-b5f6-4136dcd05ec5">

I'm not able to visit the vercel test site since I don't have access. Let me know if you're able to still reproduce the error, or if there are any other issues you notice.

Reviewer: Oops, here is the link to the Vercel site that is having the issue: [remix-cool2.vercel.app](https://remix-cool2.vercel.app/)

import React, { createContext, useContext, useRef } from "react";

type PosthogType = typeof posthog | undefined;

const PosthogContext = createContext<PosthogType>(undefined);

interface PosthogProviderProps {
  children: React.ReactNode;
}

export function PosthogProvider({ children }: PosthogProviderProps) {
  const posthogInstanceRef = useRef<PosthogType>(undefined);

  // https://react.dev/reference/react/useRef#avoiding-recreating-the-ref-contents
  // Note that in StrictMode, this will run twice.
  function getPosthogInstance() {
    if (posthogInstanceRef.current) return posthogInstanceRef.current;
    if (!window.ENV.POSTHOG_API_KEY) return undefined;

    posthogInstanceRef.current = posthog.init(window.ENV.POSTHOG_API_KEY, {
      api_host: '<ph_client_api_host>',
      person_profiles: 'identified_only'
    });
    return posthogInstanceRef.current;
  }

  return (
    <PosthogContext.Provider value={getPosthogInstance()}>
      {children}
    </PosthogContext.Provider>
  );
}

export const usePosthog = () => useContext(PosthogContext);
```

Lastly, we need to add this context to your React tree. Go to `app/entry.client.tsx` and add the following: 

> ```ts file=app/entry.client.tsx

Reviewer: Ok, checking again, the `posthog-context` seems to work without hydration errors if you use it in the `app.root.tsx` instead of `entry.client`:

```
// app/root.tsx
import { PosthogProvider } from "./contexts/posthog-context";

export default function App() {
  const { ENV } = useLoaderData<typeof loader>();

  return (
    <html lang="en">
      <head>
        <Meta />
        <Links />
      </head>
      <body>
        <PosthogProvider>
          <Outlet />
        </PosthogProvider>
        <ScrollRestoration />
        <script
          dangerouslySetInnerHTML={{
            __html: `window.ENV = ${JSON.stringify(ENV)}`,
          }}
        />
        <Scripts />
      </body>
    </html>
  );
}
```

Author: Yeah, TBH I have no idea why you're experiencing these issues on Vercel and I'm not able to replicate the hydration errors running your repo locally (whether running `npm run dev` or `npm run build && npm run start` [prod]). Happy to take your lead here and update the PR to move it to root.tsx if you can confirm that it doesn't break on your vercel deployment! 

(note that remix-cool.vercel.app right now is running the dev version of your app for some reason, and doesn't show the hydration errors)

Reviewer: None of these seem to work now 😢, let me try some more options.

Reviewer: Couldn't figure it out yet, still getting a variety of hydration errors

Reviewer: Could you publish your Remix app that works so I could test it out?

  // Note that in StrictMode, this will run twice.
  function getPosthogInstance() {
    if (posthogInstanceRef.current) return posthogInstanceRef.current;
    if (!window.ENV.POSTHOG_API_KEY) return undefined;

    posthogInstanceRef.current = posthog.init(window.ENV.POSTHOG_API_KEY, {
      // ...your_options_here
    });
    return posthogInstanceRef.current;
  }

  return (
    <PosthogContext.Provider value={getPosthogInstance()}>
      {children}
    </PosthogContext.Provider>
  );
}

export const usePosthog = () => useContext(PosthogContext);

```

Lastly, we need to add this context to your React tree. Go to `app/entry.client.tsx` and add the following: 

```ts file=app/entry.client.tsx
import { RemixBrowser } from "@remix-run/react";
import { startTransition, StrictMode, useEffect } from "react";
import { hydrateRoot } from "react-dom/client";

import { PosthogProvider } from "./contexts/posthog-context";

startTransition(() => {
  hydrateRoot(
    document,
    <StrictMode>
      <PosthogProvider>
        <RemixBrowser />
      </PosthogProvider>
    </StrictMode>,
  );
});
```

### Identifying Users 

> If your app supports identifying users, you'll want to call `posthog?.identify()` wherever you have a Distinct ID. 

Reviewer: ```suggestion
To [identify users](/docs/product-analytics/identify) call `posthog?.identify()` when you have a distinct ID. 
```
